packages/typegpu/src/resolutionCtx.ts

```diff
@@ -470,9 +464,16 @@ class ResolutionCtxImpl implements ResolutionCtx {
       }
 
       // If we got here, no item with the given slot-to-value combo exists in cache yet
-      const result = isWgslData(item)
-        ? resolveData(this, item)
-        : item.resolve(this);
+      let result: string;
+      if (isWgslData(item)) {
+        result = resolveData(this, item);
+      } else if (isDerived(item) || isSlot(item)) {
+        result = this.resolve(item);
+      } else if (isSelfResolvable(item)) {
+        result = item['~resolve'](this);
+      } else {
+        result = this.resolveValue(item);
+      }
```

Comment:
i'd go with a `const` and multiple ternary operators, but it's just a personal preference